### ISSUE TITLE
(aws_cloudfront): (cloudwatch logs access logs)

### ISSUE THREAD

**Issue by aaroncowie** [aws/aws-cdk#32279]
### Describe the feature

The cloudfront docs and the aws console refer to the current bucket based access logs as "legacy" and refer to "standard logging (v2)" instead.

I am specifically interested in the cloudwatch logs option for standard logging v2.

https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/standard-logging.html

### Use Case

Used to investigate lambda@edge integration issues.

### Acknowledgements

- [ ] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

2.170.0

### Environment details (OS name and version, etc.)

MacOS Sonoma

**Comment by khushail** [comment]
Hi @aaroncowie , your request seems to be related to AWS documentation and for any documentation changes/suggestions, there is `Feedback` button on the top right corner where you can submit feedback what you want to see.
This repository is specific to the CDK related issues. I am not sure how your request is related to CDK. Could you please verify and state your usecase.
Thanks

**Comment by aaroncowie** [comment]
My apologies, I was simply pointing out that cdk currently produces a legacy cloudfront configuration when the `logBucket` property is passed to the `Distribution` construct and does not provide a way to output the new (v2) access logging config.

The relevant sections of the cdk docs make no mention of v2
https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_cloudfront-readme.html#logging

The reason why I am interested in v2 is to make use of cloudwatch logs based logging instead of s3 based.

**Comment by jblukach** [comment]
It would be great to get this feature added to migrate away from the legacy logging option for CloudFront with CDK.

![cloudfront-settings](https://github.com/user-attachments/assets/fab567a4-f165-4b9b-a1fd-66fe49004e5a)

**Comment by khushail** [comment]
@aaroncowie ,Apologies for the delay in getting back. I don't see Cloudformation support for Cloudwatch access logs(https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-cloudfront-distribution-logging.html). 
It first needs to be provided by the cloudformation , hence I would request you to create a Request with the Cloudformation team on their [coverage roadmap](https://github.com/aws-cloudformation/cloudformation-coverage-roadmap/issues) and follow that for updates. Once that ids done, it could be made available via CDK.
Hope this would be helpful

**Comment by georeeve** [comment]
A feature request has been raised on the coverage roadmap [here](aws-cloudformation/cloudformation-coverage-roadmap#2204).

**Comment by georeeve** [comment]
After looking at the docs on how to solve this with the CloudWatch API [here](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/standard-logging.html#enable-access-logging-api), we came up with this CDK solution using the `aws-logs` L1 constructs. Note that at _least_ the `CfnDeliverySource` will need to be deployed to `us-east-1`. We have just deployed everything to that region so it's possible you might need to do extra config for cross region logs.

 ```typescript
import {
  CfnDelivery,
  CfnDeliveryDestination,
  CfnDeliverySource,
  LogGroup,
} from "aws-cdk-lib/aws-logs";

const distributionDeliverySource = new CfnDeliverySource(
  this,
  "DistributionDeliverySource",
  {
    name: "distribution-logs-source",
    logType: "ACCESS_LOGS",
    resourceArn: Stack.of(this).formatArn({
      service: "cloudfront",
      region: "",
      resource: "distribution",
      resourceName: distribution.distributionId,
    }),
  },
);

const distributionDeliveryDestination = new CfnDeliveryDestination(
  this,
  "DistributionDeliveryDestination",
  {
    name: "distribution-logs-destination",
    destinationResourceArn: new LogGroup(this, "DistributionLogGroup")
      .logGroupArn,
    outputFormat: "json",
  },
);

new CfnDelivery(this, "DistributionDelivery", {
  deliverySourceName: distributionDeliverySource.name,
  deliveryDestinationArn: distributionDeliveryDestination.attrArn,
}).node.addDependency(distributionDeliverySource);
```

**Comment by stekern** [comment]
@georeeve Thanks for the snippet! After some testing it seems like you can't get this to work without also having the log group in `us-east-1`. The [docs](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/standard-logging.html#enable-access-logging-api) can make it seem like it's possible though, but they're a bit ambigious:
> When calling the CloudWatch API to enable standard logging, you must specify the US East (N. Virginia) Region (us-east-1), even if you want to enable cross Region delivery to another destination. For example, if you want to send your access logs to an S3 bucket in the Europe (Ireland) Region (eu-west-1), use the CloudWatch API in the us-east-1 Region.

**Comment by umurkontaci** [comment]
@georeeve solid find! I went ahead made it a little construct that's a bit more L2.

You would use it like this:
```typescript
new CloudFrontLogToCloudWatch(this, 'LogDelivery', {
    distribution: this.distribution, // your cfn distribution
    outputFormat: 'json'
});
```

Here's the source:
```typescript
import {Construct} from 'constructs';
import {
    CfnDelivery,
    CfnDeliveryDestination,
    CfnDeliverySource,
    LogGroup,
} from "aws-cdk-lib/aws-logs";
import * as cloudfront from 'aws-cdk-lib/aws-cloudfront';
import {Names} from "aws-cdk-lib";

export interface CloudFrontLogToCloudWatchProps {
    distribution: cloudfront.IDistribution;
    outputFormat: 'json' | 'w3c' | 'raw' | 'plain'
}

export class CloudFrontLogToCloudWatch extends Construct {
    public readonly logGroup: LogGroup;
    constructor(scope: Construct, id: string, props: CloudFrontLogToCloudWatchProps) {
        super(scope, id);

        const distDeliverySource = new CfnDeliverySource(this, "DistributionDeliverySource", {
            name: `${Names.uniqueResourceName(this, {maxLength:55})}-src`,
            logType: 'ACCESS_LOGS',
            resourceArn: props.distribution.distributionArn
        });

        this.logGroup = new LogGroup(this, 'DistributionLogGroup', {})

        const distDeliveryDestination = new CfnDeliveryDestination(this, "DistributionDeliveryDestination", {
            name: `${Names.uniqueResourceName(this, {maxLength:55})}-dest`,
            destinationResourceArn: this.logGroup.logGroupArn,
            outputFormat: props.outputFormat
        });

        const delivery = new CfnDelivery(this, "DistributionDelivery", {
            deliverySourceName: distDeliverySource.name,
            deliveryDestinationArn: distDeliveryDestination.attrArn
        });
        delivery.node.addDependency(distDeliverySource);
        delivery.node.addDependency(distDeliveryDestination);
    }
}
```

I did not test it for cross-account (which seems like it wouldn't work without granting cross-account permissions) and I deployed this to us-east-1 (Cloudfront does everything in us-east-1 so it made sense). You don't need to mess with IAM or resource policies for in-region, same-account deployments.

**Comment by anthony-nhs** [comment]
@umurkontaci code only works if the cloudfront distribution is created in us-east-1 region.   

If you create the cloudfront distribution in another region, you must create the CfnDeliverySource, log group, CfnDeliveryDestination and  CfnDelivery in us-east-1 region

**Comment by umurkontaci** [comment]
@anthony-nhs yeah that probably makes sense. I think most of CloudFront features rely on it being in us-east-1 as it's a global service. If you were to create logs and distribution in different regions, be aware that that you might get hit with data transfer costs. I don't know if that is true for CloudFront, but it is worth confirming before getting a steep bil.

**Comment by tmokmss** [comment]
So basically, we now need at least 3 stacks to configure CloudFront from CloudFormation:

1. stack in us-east-1 for ACM certificate, WAF ACL, or Lambda@Edge
2. stack in your primary region for origin configuration
3. stack in us-east-1 for log configuration

Because stack#2 depends on stack#1, and stack#3 depends on stack#2, we cannot merge stack#1 and stack#3 despite they are in the same region. Maybe cross-region custom resources could help to reduce the complexity.

**Comment by cloudshiftchris** [comment]
I’ve solved this with a custom resource that handles the region-specific quirks.

**Comment by arditti** [comment]
@tmokmss

Can you please elaborate about the dependecy of Stack #3 on Stack #2 
I was thinking about:
Stack 1 - Creating origin on a desired region (other than us-east-1)
Stack 2 - Depend on Stack 1 (getting origin information) Creating CF/WAF/CW Logs

**Comment by tmokmss** [comment]
@arditti With that strategy, I think you could be likely to encounter circular dependencies. For example, with a Lambda function URL origin, you need an AWS::Lambda::Permission resource on stack1, which references the distribution ID on stack2. This creates a dependency from stack1 to stack2 (circular dependency). It can also lead to many cross-stack references, which often make it difficult to refactor the stacks.

Usually the best practice is to keep CF distribution and origin resources in the same stack (region).
